src/mongo_odm/query_builder.hpp

```diff
@@ -0,0 +1,473 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <cstddef>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+
+#include <bsoncxx/builder/basic/array.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/builder/basic/sub_array.hpp>
+#include <bsoncxx/builder/basic/sub_document.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+#include <bsoncxx/builder/stream/array.hpp>
+
+#define NVP(x) mongo_odm::makeNvp(&wrapBase::x, #x)
+
+#define ODM_MAKE_KEYS(Base, ...) \
+    using wrapBase = Base;       \
+    constexpr static auto fields = std::make_tuple(__VA_ARGS__);
+
+#define ODM_MAKE_KEYS_STORAGE(Base) constexpr decltype(Base::fields) Base::fields
+
+#define ODM_KEY(value) mongo_odm::hasCallIfFieldIsPresent<decltype(&value), &value>::call()
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+/**
+ * An object that represents a name-value pair of a member in an object.
+ * It is templated on the class of the member and its type.
+ */
+template <typename Base, typename T>
+struct Nvp {
+    /**
+     * Create a name-value pair from a member pointer and a name.
+     * @param  t    A pointer to the member
+     * @param  name The name of the member
+     */
+    constexpr Nvp(T Base::*t, const char *name) : t(t), name(name) {
+    }
+
+    T Base::*t;
+    const char *name;
+};
+
+/* Create a name-value pair from a object member and its name */
+template <typename Base, typename T>
+Nvp<Base, T> constexpr makeNvp(T Base::*t, const char *name) {
+    return Nvp<Base, T>(t, name);
+}
+
+/*
+ * Enumeration of possible query selectors, as well as a mapping to string values
+ */
+enum QuerySelector { QS_EQ, QS_GT, QS_GTE, QS_LT, QS_LTE, QS_NE, QS_IN, QS_NIN };
+constexpr char const *const selector_arr[] = {"$eq",  "$gt", "$gte", "$lt",
```

Comment:
The original idea was to avoid passing in invalid operators as strings when constructing the expression objects. But I don't know how much more convenient that would be in the long run, since the operator names are only referred to a handful of times in the code. I can just use string literals instead.